### PR TITLE
Droth 2439 problems with additional lanes when changing the direction main lane

### PR DESCRIPTION
**Pull request by ricardofmsimoes**
- Fix the lane creation when traffic direction change

### CODE REVIEW

File: digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/Lane.scala [hunk 1]
```diff
@@ -127,7 +127,11 @@ object LaneNumber {
     def againstDirection = 29
   }
 
-  case object Unknown extends LaneNumber { def towardsDirection = 99;  def againstDirection = 99; }
+  case object Unknown extends LaneNumber {
+    def towardsDirection = 99;
+    def againstDirection = 99;
```
Comment (marquesfm) [inline digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/Lane.scala:132]: Unnecessary semicolon on both cases

Comment (ricardofmsimoes) [reply digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/Lane.scala:132]: Fixed.

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala [hunk 2]
```diff
@@ -224,68 +289,110 @@ trait LaneOperations {
...
+            val needAdjustSideCodeLane21 = getLaneCodesForSideCodeAdjust(lanesToProcess, MainLane.againstDirection, FourthRightAdditional.againstDirection, mainLane21SideCode)
+
+            (Seq(), needAdjustSideCodeLane11 ++ needAdjustSideCodeLane21)
         }
 
         // To remove the lanes with wrong SideCode
         val toRemove = lanesToProcess.filterNot(_.id == 0L)
                                     .filter { lane =>
                                       val isLaneCodeTowards = lane.laneCode >= MainLane.towardsDirection && lane.laneCode <= FourthRightAdditional.towardsDirection
+                                      val isSideCodeTowardsOk = lane.sideCode != mainLane11SideCode.value && lane.sideCode != SideCode.BothDirections.value
```
Comment (FilPinto) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala:331]: You mean here that isSideCodeTowardsWrong, right?

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala [hunk 3]
```diff
@@ -224,68 +289,110 @@ trait LaneOperations {
...
+
                                       val isLaneCodeAgainst = lane.laneCode >= MainLane.againstDirection && lane.laneCode <= FourthRightAdditional.againstDirection
+                                      val isSideCodeAgainstOk = lane.sideCode != mainLane21SideCode.value && lane.sideCode != SideCode.BothDirections.value
```
Comment (FilPinto) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala:334]: You mean here that isSideCodeAgainstWrong, right?

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala [hunk 4]
```diff
@@ -224,68 +289,110 @@ trait LaneOperations {
...
+                                    adjustedSideCodes = changeSet.adjustedSideCodes ++ needAdjustSideCode )
+
+        (lanesToAdd, newChangeSet)
 
       case TrafficDirection.TowardsDigitizing =>
 
-        val toAdd = if (!lanesToProcess.exists(lane => lane.laneCode == mainLaneDirectionOK && lane.sideCode == SideCode.BothDirections.value))
+
+
+        val toAdd = if (!lanesToProcess.exists(lane => lane.laneCode == mainLaneDirectionOK))
```
Comment (FilPinto) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala:352]: This should be better
if(baseLane.laneCode != mainLaneDirectionOK))

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala [hunk 5]
```diff
@@ -224,68 +289,110 @@ trait LaneOperations {
...
                                     .map(_.id)
 
 
         val lanesToAdd = (lanes ++ toAdd).filterNot(lane => toRemove.contains(lane.id))
+        val newChangeSet = changeSet.copy(generatedPersistedLanes = changeSet.generatedPersistedLanes ++ lanesToAdd,
+                                    expiredLaneIds = changeSet.expiredLaneIds ++ toRemove,
+                                    adjustedSideCodes = changeSet.adjustedSideCodes ++ needAdjustSideCode )
 
-        (lanesToAdd, changeSet.copy(expiredLaneIds = changeSet.expiredLaneIds ++ toRemove, generatedPersistedLanes = changeSet.generatedPersistedLanes ++ lanesToAdd))
+        (lanesToAdd, newChangeSet)
```
Comment (FilPinto) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala:395]: this case is equal to last one, just do it like this 
      case TrafficDirection.TowardsDigitizing | TrafficDirection.AgainstDigitizing =>

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala (hunk 2 above)
Comment (ricardofmsimoes) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala:331]: done.

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala (hunk 3 above)
Comment (ricardofmsimoes) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala:334]: done.

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala (hunk 5 above)
Comment (ricardofmsimoes) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala:395]: Done

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala [hunk 6]
```diff
@@ -199,6 +226,44 @@ trait LaneOperations {
...
+    def getLaneNumbersRange(laneNumber: Int): (Int, Int) = {
+      val laneNum = LaneNumber(laneNumber)
+
+      if (laneNumber == laneNum.againstDirection)
+        (laneNum.againstDirection, FourthRightAdditional.againstDirection)
+      else if (laneNumber == laneNum.towardsDirection)
+        (laneNum.towardsDirection, FourthRightAdditional.towardsDirection)
+      else
+        (laneNum.towardsDirection, FourthRightAdditional.towardsDirection)
+    }
```
Comment (marquesfm) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala:261]: In the "else if" and "else" condition the result is the same, if this is correct we can do only like this:
```Scala
    def getLaneNumbersRange(laneNumber: Int): (Int, Int) = {
      val laneNum = LaneNumber(laneNumber)

      if (laneNumber == laneNum.againstDirection)
        (laneNum.againstDirection, FourthRightAdditional.againstDirection)
      else
        (laneNum.towardsDirection, FourthRightAdditional.towardsDirection)
    }

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala [hunk 7]
```diff
@@ -224,68 +289,110 @@ trait LaneOperations {
...
                     else
                       Seq()
 
+        val (laneCodeStart, laneCodeEnd) = getLaneNumbersRange(mainLaneDirectionOK)
+        val needAdjustSideCode = getLaneCodesForSideCodeAdjust(lanesToProcess, laneCodeStart, laneCodeEnd, SideCode.BothDirections)
+
         val toRemove = lanesToProcess.filterNot(_.id == 0L)
                                     .filter(lane => (lane.laneCode >= mainLaneDirectionRemove._1 && lane.laneCode <= mainLaneDirectionRemove._2) ||
                                               lane.sideCode != SideCode.BothDirections.value)
                                     .map(_.id)
```
Comment (marquesfm) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala:362]: Let's avoid to perform two filters on the same object, I think will be better if we do something like:
```Scala
        val toRemove = lanesToProcess.filter(lane =>
          lane.id != 0L && ((lane.laneCode >= mainLaneDirectionRemove._1 && lane.laneCode <= mainLaneDirectionRemove._2) || lane.sideCode != SideCode.BothDirections.value))
          .map(_.id)

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala [hunk 8]
```diff
@@ -224,68 +289,110 @@ trait LaneOperations {
...
```
Comment (marquesfm) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala]: Let's avoid to perform two filters on the same object, I think will be better if we do something like:
```Scala
        val toRemove = lanesToProcess.filter(lane =>
          lane.id != 0L && ((lane.laneCode >= mainLaneDirectionRemove._1 && lane.laneCode <= mainLaneDirectionRemove._2) || lane.sideCode != SideCode.BothDirections.value))
          .map(_.id)

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala (hunk 6 above)
Comment (ricardofmsimoes) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala:261]: Done.

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala (hunk 7 above)
Comment (ricardofmsimoes) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala:360]: Done

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala (hunk 8 above)
Comment (ricardofmsimoes) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala]: Done.

**Pull request merged**